fs/ref_client.go

```diff
@@ -0,0 +1,111 @@
+/*
+   Copyright The Soci Snapshotter Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package fs
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/awslabs/soci-snapshotter/soci"
+	"github.com/containerd/containerd/platforms"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactsspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/registry/remote"
+)
+
+var (
+	ErrNoReferrers = errors.New("no existing referrers")
+)
+
+/// Determines which index will be selected from a list of index descriptors
+type IndexSelectionPolicy func([]ocispec.Descriptor) (ocispec.Descriptor, error)
+
+func SelectFirstPolicy(descs []ocispec.Descriptor) (ocispec.Descriptor, error) {
+	return descs[0], nil
+}
+
+/// Responsible for making Referrers API calls to remote registry to fetch list of referrers.
+type ReferrersClient interface {
+	/// Takes in an manifest descriptor and IndexSelectionPolicy and returns a single artifact descriptor.
+	/// Returns an error (ErrNoReferrers) if the manifest descriptor has no referrers.
+	Get(context.Context, ocispec.Descriptor, IndexSelectionPolicy) (ocispec.Descriptor, error)
+}
+
+type orasReferrersClient struct {
```

Comment:
This abstraction feels weird where we create a new client that wraps the old client just to make index manifest selection a little cleaner. 

If we want this to be a wrapper client, I would prefer we embed the original `*remote.Repository` in the wrapper and then use the wrapper everywhere.

Reply:
Created a new `orasClient` struct that embeds `*remote.Repository`.